경북대 BE_탁정민 3주차 과제 (3단계)

xkrwjdals opened the pull request:
상세한 피드백 늘 감사합니다!

# 2단계 피드백 반영 사항
2단계 과제 질문1 에 대한 답변처럼 테스트 코드의 목적은 이해가 되지만 Repository Test의 경우에 JpaRepository를 extends한 interface이므로 대부분의 메서드를 제가 직접 구현하는 것이 아닌 이미 구현된 메서드를 불러와 사용하는거라 굳이 테스트 코드를 작성해야할 필요가 있을까? 하는 의문점이 생겼습니다.(어차피 JpaRepository의 메서드를 가져와서 사용하니까 Test 코드 작성에 문제만 없다면 무조건 통과하지 않나? 이런 생각이였습니다.)

그리고 질문3 에 대한 답변을 생각해보면서 `WishListController`의 memberId를 3곳에서 받는 것을 줄여보았고, API URI도 수정해보았습니다!

# 3단계 과제 질문
현재 `Pagable`과 `Page`를 사용하여 페이지네이션을 구현하였는데, 만약 thymeleaf를 사용한다면 어떻게 활용할 수 있나요?

wotjd243 committed:
Initial commit

wotjd243 committed:
feat: set up the project

xkrwjdals committed:
경북대 BE_탁정민 2주차 과제(0단계) (#37)

1주차에서 반영하지 못한 건들은 2주차 미션하면서 반영해주시면 됩니다! :)

xkrwjdals committed:
경북대 BE_탁정민 2주차 과제 (1단계) (#169)

2주차 1단계 미션하시느라 고생하셨습니다 :)
유효성 검증 로직 잘 반영하셨어요!
1단계는 머지하겠습니다. 피드백 드린것은 다음 2단계에 반영 부탁드립니다!
질문은 따로 코멘트 남겨 놓겠습니다.

xkrwjdals committed:
경북대 BE_탁정민 2주차 과제 (2단계 + 3단계) (#293)

2주차 미션 하시느라 고생 많으셨습니다! 피드백 꼼꼼히 반영해주셔서 감사합니다 :)
피드백 반영해주신건 3주차 미션 때 더 살펴볼게요! 일단 3주차 미션 시작을 위해 머지하겠습니다! 👍
질문 주셨던 것은 코멘트로 달아놨습니다. 확인해보시고 이해가 잘 안되시거나 추가 설명이 필요하시면 말씀해주세요 :) 3주차도 화이팅 입니다!

xkrwjdals committed:
fix(merge): Resolve merge conflicts

xkrwjdals committed:
docs: Update README.md

xkrwjdals committed:
refactor: Refactoring with Product, ProductRepository, ProductService to use JPA

- Implement dependency about JPA
- Delete data.sql and product table in schema.sql to use JPA not Jdbc Template

xkrwjdals committed:
refactor: Refactoring with Member, MemberRepository, MemberService to use JPA

- Delete members table in schema.sql to use JPA not Jdbc Template
- Apply Google code style to Product domain codes and optimize import

xkrwjdals committed:
fix(entity): Add protected access modifier in Member Entity

xkrwjdals committed:
refactor: Refactoring with WishList, WishListRepository, WishListService to use JPA

- Delete schema.sql to use JPA not Jdbc Template
- Apply Google code style to Product domain codes and optimize import

xkrwjdals committed:
refactor: Refactoring with WishList, WishListRepository, WishListService to use JPA

- Delete schema.sql to use JPA not Jdbc Template
- Apply Google code style to Product domain codes and optimize import

xkrwjdals committed:
feat(test): Add WishListRepositoryTest for DataJpaTest

xkrwjdals committed:
docs: Update README.md

xkrwjdals committed:
feat: Implement DataJpaTest for member

xkrwjdals committed:
fix: Change jwtUtil from controller to service

xkrwjdals committed:
fix: Change AuthInfo return type from String to Long

- Apply to AuthArgumentResolver and WishListController

xkrwjdals committed:
refactor: Add repository package

xkrwjdals committed:
refactor: Modify return value, return types

xkrwjdals committed:
docs: Update README.md

xkrwjdals committed:
feat: Add one-to-many relationship betwewen WishList and Product and WishList and Member

- Modify method on WishListService applying one-to-many relationship
- Implement findByName method on ProductRepository

xkrwjdals committed:
feat: Implement enum ErrorCode and modify RepositoryException

- Apply ErrorCode to Member, Product, WishListService

xkrwjdals committed:
feat(test): Implement DataJpaTest for Product

xkrwjdals committed:
refactor: Apply RESTful API URI on WishListController

- Modify return type of createMember and createWishList method in Service layer
- Modify Controller too

xkrwjdals committed:
refactor: Modify return type of createProduct method

- Modify Controller too

xkrwjdals committed:
refactor: Modify WishListRepositoryTest

xkrwjdals committed:
fix: Resolve merge conflicts

xkrwjdals committed:
docs: Update README.md

xkrwjdals committed:
feat: Apply Pageable and Page to WishListService and WishListController

xkrwjdals committed:
feat: Apply Pageable and Page to ProductService and ProductController

xkrwjdals committed:
feat: Modify ErrorCode Correctly

xkrwjdals committed:
fix: Apply MemberDTO return type and modified ErrorCode

xkrwjdals committed:
refactor(model): Modify WishListDTO parameters

- Delete String email in WishListDTO
- Change String ProductName to long productId in WishListDTO
- Apply to Product and WishList

xkrwjdals committed:
refactor: Delete memberId of RequestParam and if statements in Controller

- Apply it on WishListService

xkrwjdals committed:
refactor: Modify RepositoryTest correctly

xkrwjdals committed:
docs: Update README.md

jadie0319 commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -26,9 +28,13 @@ public ProductController(ProductService productService) {
+        @RequestParam(defaultValue = "0") int page, @RequestParam(defaultValue = "10") int size) {
+        Page<ProductDTO> productsPage = productService.getAllProduct(page, size);
+        HttpHeaders headers = new HttpHeaders();
+        headers.add("X-Page-Number", String.valueOf(productsPage.getNumber()));
+        headers.add("X-Page-Size", String.valueOf(productsPage.getSize()));
```
왜 헤더로 반환하는건지 이유를 알 수 있을까요?

jadie0319 commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -24,7 +28,25 @@ public ProductDTO createProduct(ProductDTO productDTO) {
-    public List<ProductDTO> getAllProduct() {
+    public Page<ProductDTO> getAllProduct(int page, int size) {
+        Pageable pageRequest = createPageRequestUsing(page, size);
+        List<Product> allProducts = productRepository.findAll();
```
findAll 의 반환값을 Page<Product> 로 받을 수 있어요.
Page 인터페이스 메서드에 대해 찾아보시면 아래 로직을 정리할 수 있을거에요~

그리고 Paging 처리 요청에 대한 응답으로 어떤 데이터가 포함되어야 할까요?
응답값으로 받은 Page 인터페이스가 제공하는 메서드를 확인하시면서 고민해보면 좋을 것 같아요!

jadie0319 commented on src/main/java/gift/service/WishListService.java:
```diff
@@ -46,42 +50,53 @@ public List<WishListDTO> getAllWishList() {
-        List<WishList> wishlists = wishListRepository.findWishListByMemberId(memberId);
-        return wishlists.stream()
+    public Page<WishListDTO> getWishListByMemberId(long memberId, int page, int size) {
+        Pageable pageRequest = createPageRequestUsing(page, size);
+
+        List<WishList> allWishlists = wishListRepository.findWishListByMemberId(memberId);
+        List<WishListDTO> allWishListDTO = allWishlists.stream()
             .map(this::convertToDTO)
             .collect(Collectors.toList());
+
+        int start = (int) pageRequest.getOffset();
+        int end = Math.min((start + pageRequest.getPageSize()), allWishlists.size());
+
+        List<WishListDTO> pageContent = allWishListDTO.subList(start, end);
```
ProductService 에 있는 피드백을 참고해서 변경해보시면 될 것 같아요~

jadie0319 commented on src/test/java/gift/MemberRepositoryTest.java:
```diff
@@ -14,18 +17,28 @@ public class MemberRepositoryTest {
     @Autowired
     private MemberRepository members;
 
+    private Member createMember(String email, String password) {
+        return new Member(email, password);
+    }
```
이 메서드는 맨 아래 위치시는케 좋을 것 같아요
테스트 클래스의 목적은 테스트 코드예요.
테스트 코드를 위에 위치시키면 코드를 읽는 사람(동료) 이 더 편하게 느낄 것 같아요~

jadie0319 commented on src/test/java/gift/ProductRepositoryTest.java:
ProductService 에서 사용중인 메서드들을 테스트 해야 할 것 같은데 어떻게 생각하시나요?

jadie0319 commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -21,9 +21,10 @@ public MemberDTO createMember(MemberDTO memberDTO) {
-    public Member findMemberByCredentials(String email, String password) {
-        return memberRepository.findByEmailAndPassword(email, password)
-            .orElseThrow(() -> new RepositoryException(ErrorCode.MEMBER_NOT_FOUND, email));
+    public MemberDTO findMemberByCredentials(String email, String password) {
+        Member foundMember = memberRepository.findByEmailAndPassword(email, password)
+            .orElseThrow(() -> new RepositoryException(ErrorCode.MEMBER_NOT_FOUND, email, password));
+        return convertToDTO(foundMember);
```
DTO 로 반환 좋습니다! 👍

jadie0319 reviewed (changes requested):
3주차 3단계 미션하시느라 고생하셨습니다!
지난 단계 피드백 잘 반영하셨어요! :)
피드백 남겨드렸으니 확인 부탁드려요~
질문 주신건 코멘트로 남겨놓겠습니다!

jadie0319 commented:
안녕하세요 정민님! 3주차 3단계 미션하시느라 고생하셨습니다.
피드백 잘 반영해주셔서 감사합니다 ㅎㅎ
좀더 보완했으면 하는 부분이 있어서 피드백 남겨놨습니다! 확인 부탁드려요!
질문해주셨던것 답변 드리겠습니다.

### 질문1
- 2단계 과제 질문1 에 대한 답변처럼 테스트 코드의 목적은 이해가 되지만 Repository Test의 경우에 JpaRepository를 extends한 interface이므로 대부분의 메서드를 제가 직접 구현하는 것이 아닌 이미 구현된 메서드를 불러와 사용하는거라 굳이 테스트 코드를 작성해야할 필요가 있을까? 하는 의문점이 생겼습니다.(어차피 JpaRepository의 메서드를 가져와서 사용하니까 Test 코드 작성에 문제만 없다면 무조건 통과하지 않나? 이런 생각이였습니다.)

저도 같은 의문을 가졌던 적이 있어서 뭔가 반갑네요 🙂 이미 구현되어 있는 메서드지만 잘 사용하고 있는지를 알기 위해서 테스트 하는 것도 있습니다. 저도 JPA 를 사용할 때 헷갈리는 부분이 있거나 확인해보고 싶은 로직이 있으면 DataJpaTest 를 만들어 확인할 때가 있어요. 엔티티 객체에서 설정한 연관관계 매핑은 잘 됐는지, Eager 또는 Lazy 로딩 설정이 의도한대로 잘 동작하는지, Cascade 는 잘 되는지 등등 여러가지를 테스트 코드를 통해 확인해볼 수 있어요. 

 물론 실제 기능을 동작시키면서 발생한 쿼리를 눈으로 확인하며 테스트 할 수 있지만 테스트 코드를 만들어두면 좀 더 빠르게 테스트할 수 있어요. 테스트 코드 실행시 발생하는 쿼리를 보며 N+1 문제는 발생하는지 등도 확인해볼 수 있어요. 

 `@SpringBootTest` 를 사용해서 서비스 로직과 Repository 를 같이 테스트 해도 되는거 아닌가? 하는 생각을 하고 계실 수도 있겠네요. 그러면 `@SpringBootTest` 와 `@DataJpaTest` 의 차이에 대해 찾아보시고 잘 모르시겠으면 질문 주세요 🙂

### 질문 2
- 현재 Pagable과 Page를 사용하여 페이지네이션을 구현하였는데, 만약 thymeleaf를 사용한다면 어떻게 활용할 수 있나요?

Paging 을 적용하셨으면 Service 메서드에서 paging 과 관련된 정보(페이지수, 전체 리스트 개수, 현재 페이지 개수 등) 와 데이터(ex: 상품목록)  를 반환해줄 수 있어요. 이렇게 Service 로 부터 얻은 정보를 API 응답값으로 반환해주고 thymeleaf 에서 해당 객체를 사용하면 되지 않을까요?  해당 데이터를 가지고 화면에 어떻게 뿌려줄 것인지는 구글링하면 관련 html 코드 예제 많이 나올 것 같아요 ㅎ

xkrwjdals committed:
refactor: Modify test code for applying feedback

jadie0319 commented on src/main/java/gift/controller/ProductController.java:
```diff
@@ -26,9 +28,13 @@ public ProductController(ProductService productService) {
-        List<ProductDTO> products = productService.getAllProduct();
-        return ResponseEntity.ok(products);
+    public ResponseEntity<Page<ProductDTO>> getAllProduct(
+        @RequestParam(defaultValue = "0") int page, @RequestParam(defaultValue = "10") int size) {
+        Page<ProductDTO> productsPage = productService.getAllProduct(page, size);
```
Pageable 객체를 요청객체로 받으면 어떨까요?
`@PageableDefault` 어노테이션 사용과 함께 고민해보시면 좋을 것 같아요!

xkrwjdals committed:
feat: Implement ProductPageDTO within pageNumber, pageSize, totalElements, List of ProductDTO

- Implement findAll method to return Page type
- Apply ProductPageDTO to Controller, Service

xkrwjdals committed:
refactor: Modify code style and name of variable

xkrwjdals committed:
feat: Implement WishListPageDTO within pageNumber, pageSize, totalElements, List of WishListDTO

- Implement findAll method to return Page type
- Apply WishListPageDTO to Controller, Service

xkrwjdals committed:
refactor: Modify Repository Test with pagination

jadie0319 commented on src/main/java/gift/service/WishListService.java:
```diff
@@ -28,60 +30,57 @@ public WishListService(WishListRepository wishListRepository,
+        Product product = productRepository.findById(wishListDTO.productId())
             .orElseThrow(() -> new RepositoryException(ErrorCode.PRODUCT_NOT_FOUND,
-                wishListDTO.productName()));
+                wishListDTO.productId()));
         Member member = memberRepository.findById(wishListDTO.memberId())
-            .orElseThrow(() -> new RepositoryException(ErrorCode.MEMBER_NOT_FOUND,
+            .orElseThrow(() -> new RepositoryException(ErrorCode.MEMBER_ID_NOT_FOUND,
                 wishListDTO.memberId()));
-        WishList wishList = new WishList(wishListDTO.email(), member,
+        WishList wishList = new WishList(member,
             product, wishListDTO.quantity());
         return convertToDTO(wishListRepository.save(wishList));
```
위시리스트를 만들때 
이미 기존에 등록되어 있던 위시리스트가 있으면 기존 위시리스트에 수량을 더해줘야 할까요?
아니면 새로운 위시리스트를 만들어야 할까요? 현재는 새로운 위시리스트를 만들고 있네요. :)

![image](https://github.com/user-attachments/assets/14e005fd-39cf-4b4a-8472-4ef16b7821c9)

xkrwjdals replied:
새로운 위시 리스트를 만드는 것이 아닌 기존 위시리스트에 수량을 더해주거나 중복되는 "memberId"와 "productId"를 갖는 위시 리스트가 존재한다고 사용자에게 알려주는 것이 맞는 것 같습니다.

그래서 `WishList` Entity에 `@UniqueConstraint`를 사용하여 중복을 방지하고, 중복되는 경우에 `GlobalExceptionHandler`를 사용하여 사용자에게 알 수 있도록 했습니다.

그런데 여기서 질문이 있습니다. 
1. 제가 구현한 것처럼 중복을 방지하고, `PUT` 요청을 통해서만 위시 리스트의 수량을 조절할 수 있게 하는 것이 좋을까요? 아니면 `POST` 요청 시 기존에 등록되어 있던 위시 리스트가 있다면 기존 위시 리스트에 수량을 더해주는 것이 좋을까요? 이는 단순히 선호도 차이일까요?
2. `POST` 요청과 `PUT` 요청 모두 `JpaRepository`의 `save()` 메서드를 사용합니다. 그래서 `@UniqueConstraint` 를 사용하여 중복을 방지하면, `PUT` 요청에서도 예외가 발생할 수도 있겠다고 생각했는데, `save()` 메서드는 내부 구현 코드를 보면 'Entity가 새로운 Entity라면 `persist()` 하고, 이미 존재하는 Entity라면 `merge()`한다 '로 동작하기 때문에 예외가 발생하지 않는다고 이해하면 될까요? (제가 이해한 것이 맞는지 궁금합니다)
3. 1번 질문에서 만약 `POST` 요청 시 기존에 등록되어 있던 위시 리스트가 있다면 기존 위시 리스트에 수량을 더해주기 위해 다음과 같은 코드를 작성하면 될까요?
```java
public WishListDTO createWishList(WishListDTO wishListDTO) {
        Product product = productRepository.findById(wishListDTO.productId())
            .orElseThrow(() -> new RepositoryException(ErrorCode.PRODUCT_NOT_FOUND,
                wishListDTO.productId()));
        Member member = memberRepository.findById(wishListDTO.memberId())
            .orElseThrow(() -> new RepositoryException(ErrorCode.MEMBER_ID_NOT_FOUND,
                wishListDTO.memberId()));
        Optional<WishList> wishList = wishListRepository.findByMemberIdAndProductId(wishListDTO.memberId(),
            wishListDTO.productId());
        if (wishList.isEmpty()) {
            WishList newWishList = new WishList(member, product, wishListDTO.quantity());
            return convertToDTO(wishListRepository.save(newWishList));
        }
        long quantity = wishList.get().getQuantity() + 1;
        WishListDTO newWishListDTO = new WishListDTO(wishListDTO.memberId(), wishListDTO.productId(), quantity);
        return updateWishListQuantity(newWishListDTO);
}
```

jadie0319 replied:
지난 미션에서 위시리스트를 추가할 수 있다 라는 요구사항만 존재하고 있어요. 구체적으로 어떻게 구현할지는 정해진게 없으니  정민님이 정해서 진행하시면 될 것 같아요. 지금은 동일한 상품을 여러개 등록할 수 있는데, 만약 등록하게되면 위시리스트 수량 변경시 예외가 발생하기 때문에 구현 방향에 대한 정리가 필요해 보여서 피드백 드렸습니다.

1.단순히 선호도 차이는 아닙니다. 요구사항에 따라 다르지만, 말씀해주신 두 방법중에 반드시 한가지의 방법만을 통해 구현해야만 하는 경우도 있어요.  그러나 지금은 요구사항이 명확하지 않으니 적절한 방법으로 구현하면 됩니다. 말씀해주신 두가지 방법은 변경로직을 엄격하게 PUT 에서만 허용할 것인지 아닌지를 판단하는게 관건이겠네요.  정민님의 방법을 찾아보세요! 일단은 정민님께서 구현하기 편한 방향으로 가도 좋을 것 같아요!

아래는 어떤 피드백을 드려야 하나 고민하다가 그냥 의식의 흐름대로 적다보니 생긴건데... 도움이 되실까 하여 남겨놨습니다. 

중복을 방지하고 PUT 요청으로만 수량이 조절 가능한 경우
- API 호출자가 동일한 상품으로 위시리스트 생성 요청(POST) 호출 -> 실패
- 중복 저장이 안된다는 응답을 받고 PUT 메서드로 다시 호출 -> 성공
- API 호출 한번으로 해결하려면 API 호출자가 미리 위시리스트가 존재하는지에 대한 여부를 알고 있어야함.

POST 시 기존에 등록된 위시리스트가 있다면 수량을 더해주게한 경우
- API 호출자가 POST 를 호출해도 수량 변경을 해준다.
- 수량을 변경하기 위해 PUT 메서드가 아닌 POST 메서드를 호출하는 경우가 발생할 수 있다.

2.save() 메서드는 Entity 의 `@Id` 를 기준으로 새 엔티티인지 아닌지를 판단하는 걸로 알고 있어요. 이건 `@UniqueConstraint` 를 어떻게 사용하셨는지 코드를 봐야 알 수 있을 것 같아요. 혹시 위시리스트의 productId 를 유니크한 키로 만들었는데 save() 메서드를 통해 동일 상품의 위시리스트가 여러개 저장이 잘 되던가요?

3.아마 그때가서 직접 코드를 봐야 알겠지만 비슷한 형태의 코드가 나오지 않을까 싶네요 ㅎ 다음주차때 배우는 내용을 
```java
여기서 1이 아니라 wishListDto 의 quantity 가 와야 할 것 같아요.
 long quantity = wishList.get().getQuantity() + 1; 
```

xkrwjdals replied:
생각해보니 PR merge된 이후에는 제가 fork한 repository에 수정된 코드를 push를 해도 확인하실 수가 없네요! 깜빡했습니다.

**중복을 방지하고 PUT 요청으로만 수량이 조절 가능한 경우**를 상정하고 코드를 작성했습니다!

`WishList` 에 `@UniqueConstraint`를 다음과 같이 사용했습니다.
```java
@Entity
@Table(name = "wishlists", uniqueConstraints = {@UniqueConstraint(columnNames = {"member_id", "product_id"})})
public class WishList {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;
    @ManyToOne
    @JoinColumn(name = "member_id", nullable = false)
    private Member member;
    @ManyToOne
    @JoinColumn(name = "product_id", nullable = false)
    private Product product;
    @Column(name = "quantity", nullable = false)
    private long quantity;

    protected WishList() {
    }
    ....
}
``` 
아래는 `GlobalExceptionHandler`에 추가한 코드입니다.
```java
    @ExceptionHandler(DataIntegrityViolationException.class)
    @ResponseStatus(HttpStatus.CONFLICT)
    public ResponseEntity<String> handleUniqueConstraint(DataIntegrityViolationException ex) {
        String message = "중복된 Member Id와 Product Id가 이미 위시 리스트가 존재합니다.";
        return ResponseEntity.status(HttpStatus.CONFLICT).body(message);
    }
``` 
`WishListService`에는 변동사항이 없습니다. 위와 같은 방법을 적용했더니 중복되는 'productId'를 가진 `WishListDTO`를 `POST` 요청으로 보냈을 때, `GlobalExceptionHandler`에 정의된 에러 메시지를 리턴합니다! 해당 방법은 구글링을 통해 알아낸 방법으로 적용했더니 잘 동작했습니다!

jadie0319 replied:
아~ unique 인덱스를 추가하는거군요? 중복저장 방어 로직 잘 짜셨네요 👍 
코드 반영해주시면 4주차 미션때 코드 확인해보겠습니다 :)

jadie0319 commented on src/main/java/gift/service/WishListService.java:
```diff
@@ -28,60 +30,57 @@ public WishListService(WishListRepository wishListRepository,
             .orElseThrow(() -> new RepositoryException(
-                ErrorCode.WISHLIST_NOT_FOUND, wishListDTO.memberId(), wishListDTO.productName()));
+                ErrorCode.WISHLIST_NOT_FOUND, wishListDTO.memberId(), wishListDTO.productId()));
 
-        Product product = productRepository.findByName(wishListDTO.productName()).
+        Product product = productRepository.findById(wishListDTO.productId()).
             orElseThrow(() -> new RepositoryException(ErrorCode.PRODUCT_NOT_FOUND,
-                wishListDTO.productName()));
+                wishListDTO.productId()));
         Member member = memberRepository.findById(wishListDTO.memberId())
-            .orElseThrow(() -> new RepositoryException(ErrorCode.MEMBER_NOT_FOUND,
+            .orElseThrow(() -> new RepositoryException(ErrorCode.MEMBER_ID_NOT_FOUND,
                 wishListDTO.memberId()));
-        WishList newWishList = new WishList(currentWishList.getId(), currentWishList.getEmail(),
-            member, product, wishListDTO.quantity());
+        WishList newWishList = new WishList(currentWishList.getId(), member, product,
+            wishListDTO.quantity());
```
1.위시리스트를 생성로직에 드린 피드백과 관련하여 에러가 발생하네요. 
56 Line 에서 위시리스트를 조회시 위시리스트가 여러개 있으면 에러가 발생하네요. 
생성로직과 관련하여 같이 고민해보시면 좋을것 같아요~

2.WishList 를 조회할 때 memberId, ProductId 로 조회를 하고 있어요.
memberId, productId 로 wishlist 를 찾을 수 없으면 예외를 발생시키죠.
그런데 62,65 Line 에서 다시 상품과 회원을 조회하는데 이 로직이 꼭 필요할까요?

xkrwjdals replied:
62~65 Line의 상품과 회원을 조회하는 조회하는 로직은 필요없을 것 같습니다!

jadie0319 commented on src/main/java/gift/service/WishListService.java:
```diff
@@ -28,60 +30,57 @@ public WishListService(WishListRepository wishListRepository,
-            .collect(Collectors.toList());
-    }
+            .stream()
+            .toList();
```
toList 사용 좋습니다 👍

jadie0319 reviewed (approved):
피드백 잘 반영해주셨습니다!
추가로 드린 피드백은 다음 주차 미션에서 반영해주시면 됩니다!
3주차 3단계 미션은 머지하겠습니다!
3주차 미션하시고 피드백 반영하시느라 고생하셨어요 👍
